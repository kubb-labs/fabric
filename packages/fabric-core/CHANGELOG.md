# @kubb/fabric-core

## 0.2.7

## 0.2.6

## 0.2.5

## 0.2.4

## 0.2.3

### Patch Changes

- [`be05a65`](https://github.com/kubb-labs/fabric/commit/be05a65f8bfa2b50ebaa2d560871949b6b8b81d6) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - Fabric export

## 0.2.2

## 0.2.1

### Patch Changes

- [`7969e4c`](https://github.com/kubb-labs/fabric/commit/7969e4ca0d8d50bd900cb5f1534d45d83a4a619a) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - rename app to fabric

## 0.2.0

### Minor Changes

- [`9a01dfd`](https://github.com/kubb-labs/fabric/commit/9a01dfd5f8941ec1d26832526f3c24460c496a48) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - Prepare v0.1.0

## 0.1.8

### Patch Changes

- [`71622a1`](https://github.com/kubb-labs/fabric/commit/71622a1986169a73b7f10d83941ffb03f81490ef) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - graphPlugin

## 0.1.7

### Patch Changes

- [`5c125c8`](https://github.com/kubb-labs/fabric/commit/5c125c8a5301616fb1aa8f7a38d93f7d55ff1849) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - update packages + progressPlugin

## 0.1.6

### Patch Changes

- [#12](https://github.com/kubb-labs/fabric/pull/12) [`ead05cc`](https://github.com/kubb-labs/fabric/commit/ead05cc1d1a57bec1d72d4159f6fcb54371bfd0c) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - - Options in install and use could be optional or required

- [`67295d8`](https://github.com/kubb-labs/fabric/commit/67295d8c210d768d1e4307c1c7a683ebe978f145) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - `fileProcessor.parse` to use parsers

- [#12](https://github.com/kubb-labs/fabric/pull/12) [`ead05cc`](https://github.com/kubb-labs/fabric/commit/ead05cc1d1a57bec1d72d4159f6fcb54371bfd0c) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - Support for barrelPlugin to create barrel files

## 0.1.5

### Patch Changes

- [`c7f32d4`](https://github.com/kubb-labs/fabric/commit/c7f32d470ae88c0667356c4f788d3292ad5f5410) Thanks [@stijnvanhullem](https://github.com/stijnvanhullem)! - Correct use declare global with Kubb.App interface

## 0.1.4

### Patch Changes

- [`2434a5a`](https://github.com/kubb-labs/fabric/commit/2434a5a1aff83672d51efab6d9598b02b5dbe635) Thanks [@stijnvanhullem](https://github.com/stijnvanhullem)! - No defaultParsers + reactPlugin to use React as compiler

## 0.1.3

### Patch Changes

- [`502cb7a`](https://github.com/kubb-labs/fabric/commit/502cb7a2d28074c2433ec3add94a07bcee86a4de) Thanks [@stijnvanhullem](https://github.com/stijnvanhullem)! - Correct parse with difference between extName for imports and extName for parser

## 0.1.2

### Patch Changes

- [`a1d95ae`](https://github.com/kubb-labs/fabric/commit/a1d95ae26ffb3ea77e389509a8fae75d1ff1ddb4) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - `fileProcessor.parse` to use parsers

## 0.1.1

### Patch Changes

- [`f4f16be`](https://github.com/kubb-labs/fabric/commit/f4f16be0486133cdfb69cfa724b98c8523eb8f83) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - expose createFileParser, createFile and FileProcessor

## 0.1.0

### Minor Changes

- [`e4f8313`](https://github.com/kubb-labs/fabric/commit/e4f8313e652044df9a5f7404221d26f5333884e2) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - expose parseFile and FileManager

## 0.0.1

### Patch Changes

- [`1145527`](https://github.com/kubb-labs/fabric/commit/1145527323c10e9a066de8ec9fd79ca439963d3b) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - Add test cases for Fabric core and React
