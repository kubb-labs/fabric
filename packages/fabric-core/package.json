{
  "name": "@kubb/fabric-core",
  "version": "0.1.8",
  "description": "Core functionality for Kubb's plugin-based code generation system, providing the foundation for transforming OpenAPI specifications.",
  "keywords": [
    "typescript",
    "plugins",
    "plugin-system",
    "code-generator",
    "codegen",
    "openapi",
    "swagger",
    "kubb"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/kubb-labs/fabric.git",
    "directory": "packages/fabric-core"
  },
  "license": "MIT",
  "author": "stijnvanhulle",
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./parsers": {
      "import": "./dist/parsers.js",
      "require": "./dist/parsers.cjs"
    },
    "./parsers/typescript": {
      "import": "./dist/parsers/typescript.js",
      "require": "./dist/parsers/typescript.cjs"
    },
    "./plugins": {
      "import": "./dist/plugins.js",
      "require": "./dist/plugins.cjs"
    },
    "./types": {
      "import": "./dist/types.js",
      "require": "./dist/types.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.cts",
  "typesVersions": {
    "*": {
      "parsers": [
        "./dist/parsers.d.ts"
      ],
      "parsers/typescript": [
        "./dist/parsers/typescript.d.ts"
      ],
      "plugins": [
        "./dist/plugins.d.ts"
      ],
      "types": [
        "./dist/types.d.ts"
      ]
    }
  },
  "files": [
    "src",
    "dist",
    "*.d.ts",
    "*.d.cts",
    "!/**/**.test.**",
    "!/**/__tests__/**"
  ],
  "scripts": {
    "build": "tsdown",
    "clean": "npx rimraf ./dist",
    "lint": "bun biome lint .",
    "lint:fix": "bun biome lint --fix --unsafe .",
    "release": "pnpm publish --no-git-check",
    "release:canary": "bash ../../.github/canary.sh && node ../../scripts/build.js canary && pnpm publish --no-git-check",
    "start": "tsdown --watch ./src",
    "test": "vitest --passWithNoTests",
    "typecheck": "tsc -p ./tsconfig.json --noEmit --emitDeclarationOnly false"
  },
  "dependencies": {
    "cli-progress": "^3.12.0",
    "fs-extra": "^11.3.2",
    "js-runtime": "^0.0.8",
    "natural-orderby": "^5.0.0",
    "p-limit": "^7.2.0",
    "serve-handler": "^6.1.6",
    "tiny-open": "^1.2.0",
    "typescript": "5.9.3"
  },
  "devDependencies": {
    "@types/cli-progress": "^3.11.6",
    "@types/fs-extra": "^11.0.4",
    "@types/serve-handler": "^6.1.4",
    "remeda": "catalog:",
    "tsdown": "catalog:",
    "typescript": "catalog:"
  },
  "engines": {
    "node": ">=20"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  }
}
