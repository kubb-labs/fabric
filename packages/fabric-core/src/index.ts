export { createApp } from './createApp.ts'
export { defineApp } from './defineApp.ts'
export { FileManager } from './FileManager.ts'
export { parseFile } from './parsers/parser.ts'
