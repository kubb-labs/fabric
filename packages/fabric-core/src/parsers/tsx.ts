import { typeScriptParser } from './typescript.ts'
import { createParser } from './createParser.ts'

export const tsxParser = createParser({
  name: 'tsx',
  install() {},
  async print(file, options = { extname: '.tsx' }) {
    return typeScriptParser.print(file, options)
  },
})
