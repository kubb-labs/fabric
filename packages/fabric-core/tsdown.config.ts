import { defineConfig } from 'tsdown'

export default defineConfig({
  entry: {
    index: 'src/index.ts',
    'parsers/typescript': 'src/parsers/typescript.ts',
    'parsers/tsx': 'src/parsers/tsx.ts',
  },
  dts: true,
  target: 'es2019',
  format: ['esm', 'cjs'],
  platform: 'node',
  sourcemap: true,
  shims: true,
  exports: true,
})
