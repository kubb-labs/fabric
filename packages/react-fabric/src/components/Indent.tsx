import dedent from 'dedent'
import indentString from 'indent-string'
import React from 'react'

type IndentProps = {
  size?: number
  children?: React.ReactNode
}

/**
 * Indents all children by `size` spaces.
 * Collapses consecutive <br /> tags to at most 2.
 */
export function Indent({ size = 2, children }: IndentProps) {
  if (!children) return null

  const filtered = React.Children.toArray(children).filter(Boolean)
  const result: React.ReactNode[] = []

  let prevWasBr = false
  let brCount = 0

  filtered.forEach((child) => {
    if (React.isValidElement(child) && child.type === 'br') {
      if (!prevWasBr || brCount < 2) {
        result.push(child)
        brCount++
      }
      prevWasBr = true
    } else {
      prevWasBr = false
      brCount = 0
      result.push(child)
    }
  })

  return (
    <>
      {result.map((child) => {
        if (typeof child === 'string') {
          const cleaned = dedent(child)
          return <>{indentString(cleaned, size)}</>
        }
        return (
          <>
            {' '.repeat(size)}
            {child}
          </>
        )
      })}
    </>
  )
}
