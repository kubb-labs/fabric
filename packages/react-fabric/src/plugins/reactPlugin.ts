import { createPlugin } from '@kubb/fabric-core/plugins'
import { createElement, type ElementType } from 'react'
import { Runtime } from '../Runtime.tsx'

export type Options = {
  stdout?: NodeJS.WriteStream
  stdin?: NodeJS.ReadStream
  stderr?: NodeJS.WriteStream
  /**
   * Set this to true to always see the result of the render in the console(line per render)
   */
  debug?: boolean
}

type ExtendOptions = {
  render(App: ElementType): Promise<void> | void
  renderToString(App: ElementType): Promise<string> | string
  waitUntilExit(): Promise<void>
}

// biome-ignore lint/suspicious/noTsIgnore: production ready
// @ts-ignore
declare module '@kubb/fabric-core' {
  interface Fabric {
    render(App: ElementType): Promise<void> | void
    renderToString(App: ElementType): Promise<string> | string
    waitUntilExit(): Promise<void>
  }
}

declare global {
  namespace Kubb {
    interface Fabric {
      render(App: ElementType): Promise<void> | void
      renderToString(App: ElementType): Promise<string> | string
      waitUntilExit(): Promise<void>
    }
  }
}

export const reactPlugin = createPlugin<Options, ExtendOptions>({
  name: 'react',
  install() {},
  inject(app, options = {}) {
    const runtime = new Runtime({ fileManager: app.context.fileManager, ...options })

    return {
      async render(App) {
        runtime.render(createElement(App))
        await app.context.events.emit('start')
      },
      async renderToString(App) {
        await app.context.events.emit('start')
        return runtime.renderToString(createElement(App))
      },
      async waitUntilExit() {
        await runtime.waitUntilExit()

        await app.context.events.emit('end')
      },
    }
  },
})
