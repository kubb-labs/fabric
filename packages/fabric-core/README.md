<div align="center">
  <h1>Fabric Core</h1>
  <a href="https://kubb.dev" target="_blank" rel="noopener noreferrer">
    <img width="180" src="https://raw.githubusercontent.com/kubb-labs/fabric/main/assets/logo.png" alt="Kubb fabric logo">
  </a>


[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![Coverage][coverage-src]][coverage-href]
[![License][license-src]][license-href]
[![Sponsors][sponsors-src]][sponsors-href]

<h4>
    <a href="https://kubb.dev/" target="_blank">Documentation</a>
    <span> Â· </span>
      <a href="https://github.com/kubb-labs/fabric/issues/" target="_blank">Report Bug</a>
    <span> Â· </span>
      <a href="https://github.com/kubb-labs/fabric/issues/" target="_blank">Request Feature</a>
</h4>
</div>

<br />

Kubb Fabric is a language-agnostic toolkit for generating code and files using JSX and TypeScript.
It offers a lightweight layer for file generation while orchestrating the overall process of creating and managing files.

> [!WARNING]
> Fabric is under active development. Until a stable 1.0 release, minor versions may occasionally include breaking changes. Please check release notes and PR titles for breaking changes.


# Features

- ðŸŽ¨ Declarative file generation â€” Create files effortlessly using JSX or JavaScript syntax.
- ðŸ“¦ Cross-runtime support â€” Works seamlessly with Node.js and Bun.
- ðŸ§© Built-in debugging utilities â€” Simplify development and inspect generation flows with ease.
- âš¡ Fast and lightweight â€” Minimal overhead, maximum performance.

## Write a TypeScript file

Below is a minimal example showing how `createFabric` works together with plugins and parsers via `fabric.use`.

```ts
import { createFabric } from '@kubb/fabric-core'
import { fsPlugin } from '@kubb/fabric-core/plugins'
import { typescriptParser, createParser } from '@kubb/fabric-core/parsers'

const fabric = createFabric()

fabric.use(fsPlugin, {
  dryRun: false,
  onBeforeWrite: (path, data) => {
    console.log('About to write:', path)
  },
  clean: { path: './generated' },
})

fabric.use(typescriptParser)

await fabric.addFile({
  baseName: 'index.ts',
  path: './generated/index.ts',
  sources: [
    { value: 'export const x = 1', isExportable: true },
  ],
})

await fabric.write()

```

# API Reference

## Core
### `createFabric(options?): Fabric`
Returns a Fabric instance with:
- `fabric.use(pluginOrParser, ...options) => Fabric` â€” register plugins and parsers.
- `fabric.addFile(...files)` â€” queue in-memory files to generate.
- `fabric.files` â€” getter with all queued files.
- `fabric.context` â€” internal context holding events, options, FileManager, installed plugins/parsers.

### `defineFabric(instance?): () => Fabric`
Factory to create your own `createFabric` with an optional bootstrap `instance(fabric)` called on creation.

### Events (emitted by the core during processing)
- `start`
- `end`
- `render { fabric }`
- `file:add { files }`
- `write:start { files }`
- `write:end { files }`
- `file:start { file, index, total }`
- `file:end { file, index, total }`
- `process:start { files }`
- `process:progress { file, source, processed, percentage, total }`
- `process:end { files }`


## Plugins
#### `fsPlugin`
Writes files to disk on `process:progress`, supports dry runs and cleaning an output folder before writing.

```
import { fsPlugin } from '@kubb/fabric-core/plugins'
```

| Option | Type                                                                 | Default | Description                                                           |
|---|----------------------------------------------------------------------|---|-----------------------------------------------------------------------|
| dryRun | `boolean`                                                            | `false` | If true, do not write files to disk.               |
| onBeforeWrite | `(path: string, data: string \| undefined) => void \| Promise<void>` | â€” | Called right before each file write on `process:progress`.            |
| clean | `{ path: string }`                                                   | â€” | If provided, removes the directory at `path` before writing any files. |

Injected `fabric.write` options (via `fsPlugin`):

| Option | Type                             | Default | Description |
|---|----------------------------------|---|---|
| extension | `Record<Extname, Extname \| ''>` | â€” | Maps input file extensions to output extensions. When set, the matching parser (by extNames) is used. |

#### `barrelPlugin`
Generates `index.ts` barrel files per folder at `process:end`. `writeEntry` creates a single entry barrel at `root`.

```
import { barrelPlugin } from '@kubb/fabric-core/plugins'
```

| Option | Type                                       | Default | Description |
|---|--------------------------------------------|---|---|
| root | `string`                                   | â€” | Root directory to generate barrel files for. |
| mode | `'all' \| 'named' \| 'propagate' \| false` | â€” | Controls how exports are generated: all exports, only named exports, propagate (skip barrels), or disabled. |
| dryRun | `boolean`                                  | `false` | If true, computes barrels but skips writing. |

Injected `fabric.writeEntry` parameters (via `barrelPlugin`):

| Param | Type                                       | Description |
|---|--------------------------------------------|---|
| root | `string`                                   | Root directory where the entry `index.ts` should be created. |
| mode | `'all' \| 'named' \| 'propagate' \| false` | Controls which export style to use for the entry barrel. |

#### `progressPlugin`
Shows a CLI progress bar by listening to core events.

```
import { progressPlugin } from '@kubb/fabric-core/plugins'
```

| Option | Type | Default | Description                                                                             |
|---|---|---|-----------------------------------------------------------------------------------------|
| â€” | â€” | â€” | This plugin has no options, it displays a CLI progress bar by listening to core events. |


#### `graphPlugin`
Shows a graph of all files

```
import { graphPlugin } from '@kubb/fabric-core/plugins'
```

| Option | Type      | Default | Description                                   |
|--------|-----------|---------|-----------------------------------------------|
| root   | `string`  |         | Root directory where to start searching from. |
| open   | `boolean` | false   | Open a webpage with the generated graph       |


#### `reactPlugin`
Enables rendering React components to the terminal or to a string. Useful for CLI UIs and templating.

```
import { reactPlugin } from '@kubb/react-fabric/plugins'
```

| Option | Type | Default | Description |
|---|---|---|---|
| stdout | `NodeJS.WriteStream` | â€” | Optional output stream used to print the rendered content while the app is running. If set, the output is written progressively. |
| stdin | `NodeJS.ReadStream` | â€” | Optional input stream for interactive components. |
| stderr | `NodeJS.WriteStream` | â€” | Optional error output stream. |
| debug | `boolean` | â€” | When true, logs render/unmount information to the console to aid debugging. |

Injected methods (via `reactPlugin`):

| Method | Signature | Description                                                                                        |
|---|---|----------------------------------------------------------------------------------------------------|
| `render` | `(App: React.ElementType) => Promise<void> \| void` | Render a React component tree to the terminal and emit the core `start` event.                     |
| `renderToString` | `(App: React.ElementType) => Promise<string> \| string` | Render a React component tree and return the final output as a string (without writing to stdout). |
| `waitUntilExit` | `() => Promise<void>` | Wait until the rendered app exits, resolves when unmounted and emits the core `end` event.         |

#### `createPlugin`

Factory to declare a plugin that can be registered via `fabric.use`.

| Field                      | Required | Description                                                                                                                  |
|----------------------------|---|------------------------------------------------------------------------------------------------------------------------------|
| `name`                     | Yes | String identifier of your plugin.                                                                                            |
| `install(fabric, options)` | Yes | Called when the plugin is registered. You can subscribe to core events and perform side effects here.                        |
| `inject?(fabric, options)` | No | Return synchronously the runtime methods/properties to merge into `fabric` (e.g. `write`, `render`). This must not be async. |

Example:

```ts
import { createFabric } from '@kubb/fabric-core'
import { createPlugin } from '@kubb/fabric-core/plugins'

const helloPlugin = createPlugin<{ name?: string }, { sayHello: (msg?: string) => void }>({
  name: 'helloPlugin',
  install(fabric, options) {
    fabric.context.events.on('start', () => {
      console.log('Fabric started')
    })
  },
  inject(fabric, options) {
    return {
      sayHello(msg = options?.name ?? 'world') {
        console.log(`Hello ${msg}!`)
      },
    }
  },
})

const fabric = createFabric()
await fabric.use(helloPlugin, { name: 'Fabric' })
fabric.sayHello() // -> Hello Fabric!
```

## Parsers
#### `typescriptParser`

Prints TS/JS imports/exports and sources, supports extname mapping for generated import/export paths.

```
import { typescriptParser } from '@kubb/fabric-core/parsers'
```

| Option | Type | Default | Description                                                                                 |
|---|---|---|---------------------------------------------------------------------------------------------|
| file | `KubbFile.File` | -| File that will be used to be parsed.                                                        |
| extname | `string` | `'.ts'` | Extension to use when emitting import/export paths (e.g., rewrite `./file` to `./file.ts`). |

#### `tsxParser`

Delegates to `typescriptParser` with TSX printing settings.

```
import { tsxParser } from '@kubb/fabric-core/parsers'
```

| Option | Type | Default | Description |
|---|---|---|---|
| file | `KubbFile.File` | -| File that will be used to be parsed.                                                        |
| extname | `string` | `'.tsx'` | Extension to use when emitting import/export paths for TSX/JSX files. |

#### `defaultParser`

Fallback parser used when no extension mapping is provided to `fabric.write`.

```
import { defaultParser } @kubb/fabric-core/parsers`
```

| Option | Type | Default | Description                                                              |
|---|---|---|--------------------------------------------------------------------------|
| file | `KubbFile.File` | -| File that will be used to be parsed.                                                        |

#### `createParser`
Factory to declare a parser that can be registered via `fabric.use` and selected by `extNames` during `fabirc.write`.

| Field                      | Required | Description                                                                                                     |
|----------------------------|---|-----------------------------------------------------------------------------------------------------------------|
| `name`                     | Yes | String identifier of your parser.                                                                               |
| `extNames`                 | Yes | List of file extensions this parser can handle (e.g. ['.ts']). Use `undefined` for the default parser fallback. |
| `install(fabric, options)` | No | Optional setup when the parser is registered (subscribe to events, set state, etc.).                            |
| `parse(file, { extname })` | Yes | Must return the final string that will be written for the given file.                                           |

Example:

```ts
import { createFabric } from '@kubb/fabric-core'
import { createParser } from '@kubb/fabric-core/parsers'

const vueParser = createParser<{ banner?: string }>({
  name: 'vueParser',
  extNames: ['.vue'],
  async install(fabric, options) {
    // Optional setup
  },
  async parse(file, { extname }) {
    const banner = file.options?.banner ?? ''
    const sources = file.sources.map(s => s.value).join('\n')
    return `${banner}\n${sources}`
  },
})

const fabric = createFabric()
fabric.use(vueParser)
fabric.use(fsPlugin); // make it possible to write to the filesystem

fabric.write({ extension: { '.vue': '.ts' } })
```

> [!NOTE]
> - `fabric.use` accepts both plugins and parsers. The `fsPlugin` handles I/O and adds `fabric.write`. Parsers decide how files are converted to strings for specific extensions.
> - When extension mapping is provided to `fabric.write`, Fabric picks a parser whose `extNames` include the fileâ€™s extension. Otherwise, the default parser is used.


<!-- Badges -->

[npm-version-src]: https://img.shields.io/npm/v/@kubb/fabric-core?flat&colorA=18181B&colorB=f58517
[npm-version-href]: https://npmjs.com/package/@kubb/fabric-core
[npm-downloads-src]: https://img.shields.io/npm/dm/@kubb/fabric-core?flat&colorA=18181B&colorB=f58517
[npm-downloads-href]: https://npmjs.com/package/@kubb/fabric-core
[license-src]: https://img.shields.io/github/license/kubb-labs/fabric.svg?flat&colorA=18181B&colorB=f58517
[license-href]: https://github.com/kubb-labs/fabric/blob/main/LICENSE
[build-src]: https://img.shields.io/github/actions/workflow/status/kubb-labs/fabric/ci.yaml?style=flat&colorA=18181B&colorB=f58517
[build-href]: https://www.npmjs.com/package/@kubb/fabric-core
[minified-src]: https://img.shields.io/bundlephobia/min/@kubb/fabric-core?style=flat&colorA=18181B&colorB=f58517
[minified-href]: https://www.npmjs.com/package/@kubb/fabric-core
[coverage-src]: https://img.shields.io/codecov/c/github/kubb-labs/fabric?style=flat&colorA=18181B&colorB=f58517
[coverage-href]: https://www.npmjs.com/package/@kubb/fabric-core
[contributors-src]: https://img.shields.io/github/contributors/kubb-labs/fabric?style=flat&colorA=18181B&colorB=f58517&label=%20
[contributors-href]: #contributors-
[sponsors-src]: https://img.shields.io/github/sponsors/stijnvanhulle?style=flat&colorA=18181B&colorB=f58517
[sponsors-href]: https://github.com/sponsors/stijnvanhulle/
