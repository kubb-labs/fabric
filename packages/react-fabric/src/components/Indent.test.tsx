import { Indent } from './Indent.tsx'
import { reactPlugin } from '../plugins/reactPlugin.ts'
import { createApp } from '@kubb/fabric-core'

describe('<Indent/>', () => {
  test('indent string children by default size', async () => {
    const Component = () => {
      return (
        <Indent>
          {`
            line1
              line2
            line3
          `}
        </Indent>
      )
    }

    const app = createApp()
    app.use(reactPlugin)
    const output = await app.renderToString(Component)

    expect(output).toMatchSnapshot()
  })

  test('indent mixed children and collapse br elements', async () => {
    const Component = () => {
      return (
        <Indent size={4}>
          Hello
          <br />
          <br />
          <br />
          world
          <br />
          <span>!</span>
        </Indent>
      )
    }

    const app = createApp()
    app.use(reactPlugin)
    const output = await app.renderToString(Component)

    expect(output).toMatchSnapshot()
  })
})
