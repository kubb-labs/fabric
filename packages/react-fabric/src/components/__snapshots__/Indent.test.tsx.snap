// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Indent/> > indent mixed children and collapse br elements 1`] = `
"    Hello    
    
    world    
    <span>!</span>"
`;

exports[`<Indent/> > indent string children by default size 1`] = `
"  line1
    line2
  line3"
`;
